### DISCUSSION TITLE
Boundary conditions for a convection-diffusion problem

### DISCUSSION THREAD

**Discussion by Irene-Moreno** [usnistgov/fipy#877] [Q&A] [answered]
Hi all,

I am trying to model mixing within an antisolvent crystallisation system in FiPy. The idea is to simulate the flow of a layer of solvent and a layer of antisolvent through a microfluidic channel. For that, I am trying to solve the convection-diffusion equation, shown below.

![IreneMoreno_convection_diffusion_eq](https://user-images.githubusercontent.com/92305869/178549623-2ef857a9-d0b3-41f0-83fe-d9c13be217bd.png)

My first question is regarding this equation though. I saw in a relevant paper in my field the convection and diffusion terms written in different sides of the equation. However, in the convection-diffusion example provided in the FiPy website, the diffusion and convection terms are in the same side of the equation, and they are equal to 0 (https://www.ctcms.nist.gov/fipy/examples/convection/generated/examples.convection.exponential1D.mesh1D.html). I'm not sure of which one to use.

I have assumed that the system is in the steady state, and therefore there is no time derivative. However, I am not sure of whether I will be able to see the diffusion effects this way. I am for now assuming that the velocity profile is constant, and I am calculating it before solving the equation and passing it as a list to the PowerLawConvectionTerm coefficient. I'm not sure of how many points I need to specify though - if in line 53 of the attached code I write "range(ny)", the velocity of the fluid on top doesn't evaluate to zero on the wall. However, if I write "range(ny+1)", it seems to be correct. 

Before trying out this, I developed a working code that only modelled diffusion (not steady state), and I managed to get reasonable results with it. However, in the results of this code I'm pretty sure I don't even have mass conservation, which makes me think that there is something (or several things) wrong with the way I have implemented it. I think one of the problems might be the specification of the boundary conditions. At the left side of the grid, I have specified constant concentrations, and at the top and bottom (the walls of the channel) I have used zero flux (so the default ones). As for the outlet BCs, I wasn't sure of what to use. I have seen some people using 'zero diffusive flux', so I tried to go with that by writing "phi.faceGrad.constrain(((0,), (0,)), where=mesh.exteriorFaces)". I'm not sure if this is correct though.

I would really appreciate some guidance regarding the boundary conditions, as well as any other problems that you can find in the code. Thank you!

## **CODE:**
```

from fipy import CellVariable, Grid2D, GaussianNoiseVariable, PowerLawConvectionTerm, DiffusionTerm, Viewer

# DECLARE VARIABLES AND DEFINE THE MESH
nx = ny = 50
dx = dy = 0.003
D = 1.6

mesh = Grid2D(nx=nx, ny=ny, dx=dx, dy=dy)

# DEFINE THE INITIAL CONDITIONS
up_conc = 0.9999
down_conc = 0.0001
noise = GaussianNoiseVariable(mesh=mesh, mean=0.5, variance=0.01).value
for phi_value in range(int(nx*ny)):
    if phi_value < int(nx*ny/2):
        noise[phi_value] = up_conc
    else:
        noise[phi_value] = down_conc

psi = CellVariable(name=r"$\psi$", mesh=mesh, hasOld=True)
phi = CellVariable(name=r"$\phi$", mesh=mesh, value=noise, hasOld=True)

# DEFINE THE BOUNDARY CONDITIONS
left_conc = [] 
for i in range(50):
    if i < ny/2:
        left_conc.append(1.)
    else:
        left_conc.append(0.)

# For the walls of the channel (top and bottom of the mesh), the BCs are zero flux (FiPy's default)
phi.constrain(value=left_conc, where=mesh.facesLeft)  # On the inlet, we start with fixed concentrations corresponding to the two layers of fluid
phi.faceGrad.constrain(((0,), (0,)), where=mesh.exteriorFaces)  # On the outlet, we want "zero-diffusive flux"

viewer = Viewer(vars=(phi,), datamin=0., datamax=1., cmap='viridis')
viewer.plot()
initial_phi = sum(phi)  # Checking if there is mass conservation

# DEFINE THE ACTIVITY MODEL AND EQUATIONS
# Calculate the velocity profiles
h = dy*(ny/2)
H = dy*ny
L = dx*nx
AP = 2700  # mPa
mu_e = 1.14715  # mPa·s
mu_w = 0.89  # mPa·s
A1 = -(1 + (mu_e/mu_w -1)*(h**2)/(H**2))/(1+ (mu_e/mu_w -1)*h/H)

vprof_1 = []
vprof_2 = []
y = 0
for ny_val in range(ny+1):  # There are ny nodes, but if I don't write ny+1, the upper velocity profile doesn't start in 0.
    if ny_val < ny/2:
        vx1e = (AP*(H**2)/(2*mu_e*L))*(1 - y/H)*(1 + y/H + A1)
        vprof_1.append(vx1e)
    else:
        vx2w = (-AP*(H**2)/(2*mu_w*L))*(y/H)*(y/H + A1)
        vprof_2.append(vx2w)
    y += dy

vprof = []
for ny_val in range(ny):
    for nx_val in range(nx):
        if ny_val < int(ny/2):
            vprof.append(vprof_1[ny_val])
        else:
            vprof.append(vprof_2[ny_val-int(ny/2)])

velocity = CellVariable(mesh=mesh, rank=1, value=0.0)
velocity[0] = 0
velocity[1] = vprof

# Define the equation
eq = (PowerLawConvectionTerm(coeff=velocity, var=phi) == DiffusionTerm(coeff=D, var=phi))

# SOLVE THE EQUATION
eq.solve()

viewer.plot()
print(initial_phi)
print(sum(phi))

# END

```

**Comment by wd15** [comment, marked as answer]
Hi @Irene-Moreno, below is an improved version of your code. The main thing fixed here is that the loops have been removed. Working with FiPy should never require using loops. Also, I would recommend adding a transient term for debugging purposes. You can increase the time step as much as you want when things are working. I'm not sure what the boundary conditions should be from your description above. Can you write them out with mathematics? Then we can figure out if mass should be conserved. Once we figure out the correct boundary conditions then we should try solving with 0 convection and 0 diffusion to test.

```
from fipy import CellVariable, Grid2D, GaussianNoiseVariable, PowerLawConvectionTerm, DiffusionTerm, Viewer, TransientTerm

# DECLARE VARIABLES AND DEFINE THE MESH
nx = ny = 50
dx = dy = 0.003
lx = dx * nx
ly = dy * ny
D = 1.6

mesh = Grid2D(nx=nx, ny=ny, dx=dx, dy=dy)
x, y = mesh.cellCenters
X, Y = mesh.faceCenters

# DEFINE THE INITIAL CONDITIONS
up_conc = 0.9999
down_conc = 0.0001

phi = CellVariable(name=r"$\phi$", mesh=mesh, value=down_conc, hasOld=True)
phi[y < ly / 2] = up_conc

# DEFINE THE BOUNDARY CONDITIONS

phi.constrain(1. * (Y < ly / 2), where=mesh.facesLeft)
phi.faceGrad.constrain(0, where=mesh.facesRight)


viewer = Viewer(vars=(phi,), datamin=-1.5, datamax=1.5, cmap='viridis')
viewer.plot()
initial_phi = sum(phi)  # Checking if there is mass conservation

# DEFINE THE ACTIVITY MODEL AND EQUATIONS
# Calculate the velocity profiles
h = ly / 2
H = ly
L = lx
AP = 2700  # mPa
mu_e = 1.14715  # mPa·s
mu_w = 0.89  # mPa·s
A1 = -(1 + (mu_e/mu_w -1)*(h**2)/(H**2))/(1+ (mu_e/mu_w -1)*h/H)

velocity = CellVariable(mesh=mesh, rank=1, value=0.0)
velocity[1] = (-AP*(H**2)/(2*mu_w*L))*(y/H)*(y/H + A1)
velocity[1, y < ly / 2] = ((AP*(H**2)/(2*mu_e*L))*(1 - y/H)*(1 + y/H + A1))[y < ly / 2]
input('stopped')

# Define the equation
eq = TransientTerm(var=phi) + PowerLawConvectionTerm(coeff=velocity, var=phi) == DiffusionTerm(coeff=D, var=phi)

# SOLVE THE EQUATION
eq.solve(dt=1.)

viewer.plot()
print(initial_phi)
print(sum(phi))

input('stopped')
# END
```

**Reply by Irene-Moreno** [reply]
Hi @wd15,

Thank you so much for your answer! The code improvements were indeed very helpful, I didn't realise how much of it could be simplified. I took the boundary conditions from Inguva et al., 2021 (DOI: 10.1016/j.ece.2021.01.011). For the outlet, the zero-diffusive flux BCs take this shape:

![BCs](https://user-images.githubusercontent.com/92305869/180417668-64cd4ad8-1af3-42cd-a6c7-6a2d27605dc9.png)

where w and e are, respectively, the concentration of components W and E in the system. Since in my code phi corresponds with the mass fraction of component E, it should be possible to re-write them as the total concentration multiplied by the derivative of phi w.r.t. x. Also, since there is no chemical reaction or accumulation in this system, I'm not sure of why there wouldn't it be mass conservation.

Finally, I have some general questions about the code, and I would really appreciate it if you could answer them or point me to the right resources. (1) What is the difference between an outlet BC (exterior faces) and facesRight? (2) Why does a transient term help with debugging? (3) What is the syntax of the convection coefficients? I have noticed that you pass where to apply a certain value with [y < ly / 2] at the end.

Thank you again!

**Reply by wd15** [reply]
> Hi @wd15,
> 
> Thank you so much for your answer! The code improvements were indeed very helpful, I didn't realise how much of it could be simplified. I took the boundary conditions from Inguva et al., 2021 (DOI: 10.1016/j.ece.2021.01.011). For the outlet, the zero-diffusive flux BCs take this shape:
> 
> ![BCs](https://user-images.githubusercontent.com/92305869/180417668-64cd4ad8-1af3-42cd-a6c7-6a2d27605dc9.png)
> 
> where w and e are, respectively, the concentration of components W and E in the system. Since in my code phi corresponds with the mass fraction of component E, it should be possible to re-write them as the total concentration multiplied by the derivative of phi w.r.t. x. Also, since there is no chemical reaction or accumulation in this system, I'm not sure of why there wouldn't it be mass conservation.
> 

I'll can answer your questions below right now. It'll take me a bit longer to work on fixing the boundary conditions (some time next week, hoepfully).

> Finally, I have some general questions about the code, and I would really appreciate it if you could answer them or point me to the right resources.

> (1) What is the difference between an outlet BC (exterior faces) and facesRight?

Exterior faces in FiPy refers to all exterior faces or faces on the boundary. It has nothing to do with any outlet boundary conditions. I used `facesRight` as I was assuming that the faces on the right are the outlet faces (although I'm not sure if it's defined correctly currently).

> (2) Why does a transient term help with debugging?

Basically, it allows us to see the system evolve slowly instead of having a catastrophic failure that we can't understand. Moreover, a lot of systems require an evolution to reach steady state / equilibrium conditions and are unstable without time stepping. Infinite time steps just hand the job of evolving the system to the linear solver.

> (3) What is the syntax of the convection coefficients? I have noticed that you pass where to apply a certain value with [y < ly / 2] at the end.

The convection coefficient is a `CellVariable` that shares the fancy indexing capabilities that a Numpy array has. You can assign values to a Numpy array with a mask. For example, 

    >>> import numpy as np
    >>> a = np.arange(10)
    >>> a[a < 5] = 1
    >>> print(a)
    [1 1 1 1 1 5 6 7 8 9]

In our case `y < ly / 2` generates a boolean array (or `CellVarialbe` that has all the properties of a Numpy array) that we can then use as a mask to assign values to only a subset of `velocity`, which is another `CellVariable` (with the properties of a Numpy array).

Hope that helps.

**Reply by Irene-Moreno** [reply]
Hi @wd15,

Thank you for the detailed answers! They were very helpful for understanding how FiPy works. Did you have a chance to look at how to implement the boundary conditions I sent in my last message?

Thanks again,
Irene

**Reply by wd15** [reply]
Hi @Irene-Moreno, I just tried to figure out the boundary condition issue, but I don't understand the way you've written them or described them. I'm not sure what W and E are in terms of Phi. We only have one variable in the code (Phi). Can you write the boundary conditions in terms of Phi?

**Reply by Irene-Moreno** [reply]
Hi @wd15,

I'm sorry for the confusion! The equation in terms of phi takes the shape shown in the image below. In this code, phi corresponds with the mole fraction of ethanol (E) and, since it is a binary system, the mole fraction of water is (1-phi).

![CH BCs](https://user-images.githubusercontent.com/92305869/189339372-ca82b1a0-5196-49ee-9e31-85a260b67a8b.png)

I also have a question regarding the effect of the BCs. I would like to know if those boundary conditions (constant concentration) can affect the results obtained at other parts (x < L) of the map. The physical meaning of these BCs is that, at the point x=L, the concentration profile of the system will not change. However, it is not guaranteed that the mixture will be completely homogeneous when it reaches x = L. Therefore, saying that the concentration profile is constant would not make physical sense and I would like to check that the BC won't affect the general results (hopefully this situation of not simulating the whole evolution would never happen, but just making sure!).

Thank you,
Irene

**Reply by wd15** [reply]
> Hi @wd15,
> 
> I'm sorry for the confusion! The equation in terms of phi takes the shape shown in the image below. In this code, phi corresponds with the mole fraction of ethanol (E) and, since it is a binary system, the mole fraction of water is (1-phi).

Ok, then it seems like the boundary condition in the code is fine. I ran it with some loops with larger time steps and it seems to conserve mass (see code below).

> I also have a question regarding the effect of the BCs. I would like to know if those boundary conditions (constant concentration) can affect the results obtained at other parts (x < L) of the map. The physical meaning of these BCs is that, at the point x=L, the concentration profile of the system will not change. However, it is not guaranteed that the mixture will be completely homogeneous when it reaches x = L. Therefore, saying that the concentration profile is constant would not make physical sense and I would like to check that the BC won't affect the general results (hopefully this situation of not simulating the whole evolution would never happen, but just making sure!).

The way we have specified the boundary conditions simply enforces a zero flux for the diffusion term on the right hand boundary. The velocity is zero in the x-direction as we've defined it, so there is no flow in or out of the domain at all on the right hand side so the zero gradient has no impact on the convection term.

```
from fipy import CellVariable, Grid2D, GaussianNoiseVariable, PowerLawConvectionTerm, DiffusionTerm, Viewer, TransientTerm

# DECLARE VARIABLES AND DEFINE THE MESH
nx = ny = 50
dx = dy = 0.003
lx = dx * nx
ly = dy * ny
D = 1.6

mesh = Grid2D(nx=nx, ny=ny, dx=dx, dy=dy)
x, y = mesh.cellCenters
X, Y = mesh.faceCenters

# DEFINE THE INITIAL CONDITIONS
up_conc = 0.9999
down_conc = 0.0001

phi = CellVariable(name=r"$\phi$", mesh=mesh, value=down_conc, hasOld=True)
phi[y < ly / 2] = up_conc

# DEFINE THE BOUNDARY CONDITIONS

phi.constrain(1. * (Y < ly / 2), where=mesh.facesLeft)
phi.faceGrad.constrain(0, where=mesh.facesRight)


viewer = Viewer(vars=(phi,), datamin=-1.5, datamax=1.5, cmap='viridis')
viewer.plot()
initial_phi = float(sum(phi))  # Checking if there is mass conservation

# DEFINE THE ACTIVITY MODEL AND EQUATIONS
# Calculate the velocity profiles
h = ly / 2
H = ly
L = lx
AP = 2700  # mPa
mu_e = 1.14715  # mPa·s
mu_w = 0.89  # mPa·s
A1 = -(1 + (mu_e/mu_w -1)*(h**2)/(H**2))/(1+ (mu_e/mu_w -1)*h/H)

velocity = CellVariable(mesh=mesh, rank=1, value=0.0)
velocity[1] = (-AP*(H**2)/(2*mu_w*L))*(y/H)*(y/H + A1)
velocity[1, y < ly / 2] = ((AP*(H**2)/(2*mu_e*L))*(1 - y/H)*(1 + y/H + A1))[y < ly / 2]

# Define the equation
eq = TransientTerm(var=phi) + PowerLawConvectionTerm(coeff=velocity, var=phi) == DiffusionTerm(coeff=D, var=phi)

# SOLVE THE EQUATION
for _ in range(100):
    phi.updateOld()
    eq.solve(phi, dt=10.)
    viewer.plot()
    print(initial_phi)
    print(sum(phi))
    input('next step')

input('stopped')
# END
```

**Reply by Irene-Moreno** [reply]
Hi @wd15,

Thank you! Your insight in how to formulate the BCs and simplify the code was really helpful, now I have a better understanding of how to work with FiPy.

I thought that in the code above the velocity in the x direction was constant, not zero. The idea I had in mind was to specify the velocity profile only in the y direction, but letting the velocity for a particular x be the same across the mesh. If the code is not doing that as it is, how can I implement that?

Also, I'm having some problems with the value of phi. The equation I am solving defines phi as the mole fraction of ethanol, and therefore it should inherently stay within 0 and 1 (i.e. no solution obtained by the solver should yield a value outside that range). However, when I change the value of D in the code above, I get phi values much higher than 1. So there is "mass conservation", but the final solution violates physical laws (see the maps below). Do you have any idea of why this could be happening? The equation is a standard diffusion-convection problem, so I don't think it's that, and the BCs used should be correct too. 

![image](https://user-images.githubusercontent.com/92305869/192832726-e9bdca2d-1aa5-4512-8123-01f783239fdc.png) ![image](https://user-images.githubusercontent.com/92305869/192832657-04f92524-07f1-4e0d-9d44-035a3f28da63.png)

In the maps above (left: initial conditions, right: final solution), I deleted the minimum and maximum value specifications for the viewer.

The only thing I have changed from the code you sent me is the order of the equations of the velocity profile, as water should be at the top of the channel, and the sign of the convection term, since I had an error there (before it was positive, and now I changed it to negative in accordance with the eq shown in this tutorial: https://www.ctcms.nist.gov/fipy/examples/convection/generated/examples.convection.exponential1D.mesh1D.html). 

## CODE:
```
from fipy import CellVariable, Grid2D, GaussianNoiseVariable, PowerLawConvectionTerm, DiffusionTerm, Viewer, TransientTerm
import numpy as np


# DECLARE VARIABLES AND DEFINE THE MESH

nx = ny = 50
dx = dy = 0.003
lx = dx * nx
ly = dy * ny
D = 0.16

mesh = Grid2D(nx=nx, ny=ny, dx=dx, dy=dy)
x, y = mesh.cellCenters
X, Y = mesh.faceCenters


# DEFINE THE INITIAL CONDITIONS

up_conc = 0.9999
down_conc = 0.0001

phi = CellVariable(name=r"$\phi$", mesh=mesh, value=down_conc, hasOld=True)  # Water should be on top, ethanol at the bottom
phi[y < ly / 2] = up_conc

viewer = Viewer(vars=(phi,), cmap='viridis')
viewer.plot()


# DEFINE THE BOUNDARY CONDITIONS

phi.constrain(1. * (Y < ly / 2), where=mesh.facesLeft)
phi.faceGrad.constrain(0, where=mesh.facesRight)

viewer.plot()

initial_phi = sum(phi)  # Checking if there is mass conservation


# DEFINE THE ACTIVITY MODEL AND EQUATIONS

# Calculate the velocity profiles

h = ly / 2
H = ly
L = lx
AP = 2700  # mPa
mu_e = 1.14715  # mPa·s
mu_w = 0.89  # mPa·s

A1 = -(1 + (mu_e/mu_w -1)*(h**2)/(H**2))/(1+ (mu_e/mu_w -1)*h/H)

velocity = CellVariable(mesh=mesh, rank=1, value=0.0)
velocity[1] = (-AP*(H**2)/(2*mu_w*L))*(y/H)*(y/H + A1)
velocity[1, y > ly / 2] = ((AP*(H**2)/(2*mu_e*L))*(1 - y/H)*(1 + y/H + A1))[y > ly / 2]

input('stopped')

# Define the equation

eq = TransientTerm(var=phi) - PowerLawConvectionTerm(coeff=velocity, var=phi) == DiffusionTerm(coeff=D, var=phi)


# SOLVE THE EQUATION

for _ in range(100):
    phi.updateOld()
    eq.solve(phi, dt=10.)

    viewer.plot()
    print(initial_phi)
    print(sum(phi))

print(np.array(phi))
input('stopped')
```

**Reply by guyer** [reply]
> I thought that in the code above the velocity in the x direction was constant, not zero. The idea I had in mind was to specify the velocity profile only in the y direction, but letting the velocity for a particular x be the same across the mesh. If the code is not doing that as it is, how can I implement that?

I don't see any problem with the way you've coded it. What makes you think it's not working?

> Also, I'm having some problems with the value of phi. The equation I am solving defines phi as the mole fraction of ethanol, and therefore it should inherently stay within 0 and 1 (i.e. no solution obtained by the solver should yield a value outside that range). However, when I change the value of D in the code above, I get phi values much higher than 1. So there is "mass conservation", but the final solution violates physical laws (see the maps below). Do you have any idea of why this could be happening? The equation is a standard diffusion-convection problem, so I don't think it's that, and the BCs used should be correct too.

There is nothing in your equation which captures this constraint of $0 \le \phi \le 1$. Due to the velocity field, everything "piles up" at the bottom boundary, but there's nothing to say that it shouldn't. Typically, you would need a term like $1 - \phi$ or $\phi(1-\phi)$ in your diffusion and velocity coefficients to prevent unconstrained growth of your field values. This should come from the derivation of your governing equations when the appropriate constraints are imposed.

I recommend you slow things way down, in order to better understand what's happening. I used `dt=1e-5`.

**Reply by Irene-Moreno** [reply]
Hi @guyer,

Thank you for your answer! It had completely slipped past me that the equation was wrong, and I would have lost a lot of time working on an erroneous system if you hadn't spotted it. Your comments were really useful, thank you.

I derived the equation again as you suggested, and included several new terms: firstly, Maxwell-Stefan diffusion was used to describe the gradient of phi in terms of the chemical potential, and a simplified version of the Margules activity model is added to that. In the future I want to include more terms in this equation, but I need this to work first.

![New equation](https://user-images.githubusercontent.com/92305869/195142766-164890f1-100d-41af-9b9c-d979fdb30b98.png)

However, the code still gives me an erroneous output (the map goes to 0 and the two bottom rows have weird values, this behaviour intensifies from top to bottom). This reminded me of what you said in your previous response, that the velocity field is making everything pile up at the bottom boundary. Would you mind explaining me briefly how to check that? It could help me understand why this version of the code doesn't work either.

![image](https://user-images.githubusercontent.com/92305869/195150794-f09816d9-9393-4284-a494-6d8320d4e911.png)

I am not sure of why it shows this behaviour, so I have been looking at stability conditions, in case it's that. I was reading the FiPy manual, and there are several inequalities that need to be fulfilled to get physically realistic solutions (https://www.ctcms.nist.gov/fipy/documentation/numerical/scheme.html, page 63 of chapter 11, section 5 - Numerical schemes). If you see the code below, the diffusion coefficient (gamma in those equations) depends on phi, so do I need to check those inequalities in every iteration? Also, in your experience, would you say it has to do with that, or is there something bigger I'm missing?

As for the velocity profile, I thought it was working differently to how I intended because of a previous comment by Dr Wheeler (https://github.com/usnistgov/fipy/discussions/877#discussioncomment-3611116). It's good to know the error doesn't come from there though!

I realise how much time I've taken already, so it would be perhaps easier to set up a call at a time that suits you and discuss this? If communicating through GitHub is easier for you, I'm more than happy to continue this way!

### CHANGES TO THE CODE:
```
from fipy import CellVariable, Grid2D, GaussianNoiseVariable, PowerLawConvectionTerm, ImplicitSourceTerm, DiffusionTerm, Viewer, TransientTerm
import numpy as np

# DECLARE VARIABLES AND DEFINE THE MESH

nx = ny = 50
dx = dy = 0.003
lx = dx * nx
ly = dy * ny
D = 0.16

mesh = Grid2D(nx=nx, ny=ny, dx=dx, dy=dy)
x, y = mesh.cellCenters
X, Y = mesh.faceCenters


# DEFINE THE INITIAL CONDITIONS

up_conc = 0.9999
down_conc = 0.0001

psi = CellVariable(name=r"$\psi$", mesh=mesh, hasOld=True)
phi = CellVariable(name=r"$\phi$", mesh=mesh, value=down_conc, hasOld=True)  # Water should be on top, ethanol at the bottom
phi[y < ly / 2] = up_conc

# viewer = Viewer(vars=(phi,), datamin=0., datamax=1., cmap='viridis')
viewer = Viewer(vars=(phi,), cmap='viridis')
viewer.plot()


# DEFINE THE BOUNDARY CONDITIONS

phi.constrain(1. * (Y < ly / 2), where=mesh.facesLeft)
phi.faceGrad.constrain(0, where=mesh.facesRight)

viewer.plot()

initial_phi = sum(phi)  # Checking if there is mass conservation


# DEFINE THE ACTIVITY MODEL AND EQUATIONS
# Calculate the velocity profiles

h = ly / 2
H = ly
L = lx
AP = 2700  # mPa
mu_e = 1.14715  # mPa·s
mu_w = 0.89  # mPa·s

A1 = -(1 + (mu_e/mu_w -1)*(h**2)/(H**2))/(1+ (mu_e/mu_w -1)*h/H)


velocity = CellVariable(mesh=mesh, rank=1, value=0.0)
velocity[1] = (-AP*(H**2)/(2*mu_w*L))*(y/H)*(y/H + A1)
velocity[1, y > ly / 2] = ((AP*(H**2)/(2*mu_e*L))*(1 - y/H)*(1 + y/H + A1))[y > ly / 2]

input('stopped')

# Define the equation
A = 1.19845
dfdphi = A * (1-phi)**2
d2fdphi2 = 2*A*(1-phi)

eq1 = TransientTerm(var=phi) - PowerLawConvectionTerm(coeff=velocity*phi, var=psi) == DiffusionTerm(coeff=D*phi, var=psi)
eq2 = (ImplicitSourceTerm(coeff=1., var=psi) == ImplicitSourceTerm(coeff=d2fdphi2, var=phi) - d2fdphi2 * phi + dfdphi)
eq = eq1 & eq2

# SOLVE THE EQUATION

for _ in range(500):
    res = 1.
    while res > 1.0e-5:
        res = eq.sweep(dt=1e-5)
    phi.updateOld()
    psi.updateOld()

    viewer.plot()
    print(initial_phi)
    print(sum(phi))

print(np.array(phi))
input('stopped')

```

**Comment by guyer** [comment]
A few things to look at:
- Put your terms in your equation in [canonical form for FiPy](https://www.ctcms.nist.gov/fipy/documentation/numerical/equation.html), e.g.,
    - $\vec{v}\cdot\phi\nabla\left(A\left(1-\phi\right)^2\right)$ is not of form $\nabla\cdot\left(\vec{u}\phi\right)$
    - $D\phi\nabla^2\left(A\left(1-\phi\right)^2\right)$ is not of form $\nabla\cdot\left(\Gamma\nabla\phi\right)$
- Because your terms are not in canonical form, your code does not implement the equation you wrote.
- > there are several inequalities that need to be fulfilled to get physically realistic solutions (https://www.ctcms.nist.gov/fipy/documentation/numerical/scheme.html, page 63 of chapter 11, section 5 - Numerical schemes). If you see the code below, the diffusion coefficient (gamma in those equations) depends on phi, so do I need to check those inequalities in every iteration?

    You are not responsible for checking those inequalities. By choosing different subclasses of `ConvectionTerm`, FiPy will implement different discretization schemes. Some will behave better or worse, but I don't think this is something to worry about right now. `PowerLawConvectionTerm` usually works pretty well.
- How did you choose the residual tolerance in your sweep loop of `1.0e-5`? You should choose a residual tolerance that shows that your equations have converged (the residual no longer changes very much). Until you understand the properties of your equations, you may be better off sweeping a fixed number of times; maybe ten or twenty. Don't worry about computational efficiency at this point.
- > This reminded me of what you said in your previous response, that the velocity field is making everything pile up at the bottom boundary.

  If you go back to your previous code, reduce the time step, and then either remove the convection term or the diffusion term, you can see the effect of each in isolation. Your convection term causes all of the material to flow toward the bottom of the domain. Your boundary condition doesn't let the material leave. As a result, the concentration accumulates without bound.
